### PR TITLE
JSON output

### PR DESCRIPTION
**Pull request by peterschrammel**
--json-ui option
and JSON output for
- show-properties
- show-goto-functions (currently only function list)

further TODOs:
- show-loops
- source location output

### CODE REVIEW

File: src/util/ui_message.cpp [hunk 1]
```diff
@@ -47,13 +47,19 @@ ui_message_handlert::ui_message_handlert(
...
+      std::cout << json_program;
+    }
     break;
     
   case PLAIN:
     break;
     
-  default:;
+  default:
+    assert(false);
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: My suggestion is to remove default: statements wherever possible as the compiler will warn you about missing cases.

File: src/util/ui_message.cpp [hunk 2]
```diff
@@ -289,10 +289,10 @@ void ui_message_handlert::json_ui_msg(
     result.new_element(xml(location));
   #endif
 
-  result["text"].value=msg1;
-  result["type"].value=type;
+  result["Text"] = json_stringt(msg1);
+  result["Type"] = json_stringt("Message");
+  result["MessageType"] = json_stringt(type);
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: nit picking: no space around = in assignments

File: src/util/ui_message.cpp [hunk 3]
```diff
@@ -289,10 +289,10 @@ void ui_message_handlert::json_ui_msg(
...
   
-  std::cout << result;
-  std::cout << std::endl;
+  std::cout << ",\n" << result;
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: I am worried about the leading "," that this will generate. How is that avoided?

File: src/cbmc/bmc.cpp [hunk 4]
```diff
@@ -223,6 +224,14 @@ void bmct::report_success()
       std::cout << "\n";
     }
     break;
+
+  case ui_message_handlert::JSON_UI:
+    {
+      json_objectt json_result;
+      json_result["CProverStatus"] = json_stringt("Success");
```
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: nit picking: no space around =

File: src/util/ui_message.cpp [hunk 5]
```diff
@@ -289,9 +289,8 @@ void ui_message_handlert::json_ui_msg(
     result.new_element(xml(location));
   #endif
 
-  result["Text"] = json_stringt(msg1);
-  result["Type"] = json_stringt("Message");
   result["MessageType"] = json_stringt(type);
+  result["MessageText"] = json_stringt(msg1);
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: I'd suggest merging with the previous commit

File: src/goto-programs/show_properties.cpp [hunk 6]
```diff
@@ -118,17 +124,99 @@ Function: show_properties
...
+    //const irep_idt &function=location.get_function();
+    const irep_idt &property_class=source_location.get_property_class();
+    const irep_idt description=
+      (comment==""?"assertion":comment);
+      
+    irep_idt property_id=source_location.get_property_id();
+
+    json_objectt &json_property = 
+      json_properties.push_back(jsont()).make_object();
+    json_property["Name"] = json_stringt(id2string(property_id));
```
Comment (tautschnig) [inline src/goto-programs/show_properties.cpp]: nit picking: no space around =

File: src/goto-programs/show_properties.cpp [hunk 7]
```diff
@@ -148,6 +236,11 @@ void show_properties(
...
   const namespacet ns(goto_model.symbol_table);
-  show_properties(ns, ui, goto_model.goto_functions);
+  if(ui == ui_message_handlert::JSON_UI)
+    show_properties_json(ns, goto_model.goto_functions);
+  else   
+    show_properties(ns, ui, goto_model.goto_functions);
 }
 
+
+
```
Comment (tautschnig) [inline src/goto-programs/show_properties.cpp]: Spurious blank lines at EOF added

File: src/goto-programs/show_goto_functions.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    json_result["Functions"] = json_functions;
+    std::cout << ",\n" << json_result;
+  }
+  break;
+      
+  case ui_message_handlert::PLAIN:
+    goto_functions.output(ns, std::cout);
+    break;
+
+  default:
```
Comment (tautschnig) [inline src/goto-programs/show_goto_functions.cpp]: See other comment: consider removing default: statement

File: src/goto-programs/show_goto_functions.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    assert(false);
+  }
+  }
```
Comment (tautschnig) [inline src/goto-programs/show_goto_functions.cpp]: Wrong indent to closing }

**Comment by tautschnig** [comment]
@peterschrammel It seems that pull request also needs rebasing before it can be merged? Thanks!

File: src/cbmc/bmc.cpp (hunk 4 above)
Comment (peterschrammel) [inline src/cbmc/bmc.cpp]: I'm against this convention as it reduces readability...

File: src/util/ui_message.cpp (hunk 3 above)
Comment (peterschrammel) [inline src/util/ui_message.cpp]: We have the following convention: The constructor generates the first entry without trailing comma. Every new entry produces a leading comma.

**Comment by peterschrammel** [comment]
I have taken into account all your comments now.

**Comment by tautschnig** [comment]
Thanks @peterschrammel for all the updates! github says there are still conflicts that need resolving? (Or maybe these are new conflicts.)
On a different note, the commit includes changes to remove_unused_functions that seem unrelated to the json UI?

File: src/util/ui_message.cpp (hunk 3 above)
Comment (tautschnig) [inline src/util/ui_message.cpp]: It might be worth adding a comment to explicitly write down this convention?

File: src/cbmc/bmc.cpp (hunk 4 above)
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: I'm ok either way, but I think consistency is key. That is, at least within one function or, better, one file, the same rules should be followed. I'd suggest adding a whitespace-changes-only commit doing that.

**Comment by peterschrammel** [comment]
removed remove_unused_functions changes and the remaining enum-switch default

**Comment by kroening** [comment]
Partially merged; note that show_goto_functions.cpp was missing.

**Pull request closed without merging**